compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceLowering.kt

```diff
@@ -201,223 +184,253 @@ internal class PropertyReferenceLowering(val context: JvmBackendContext) : Class
             isStatic = true
             visibility = JavaDescriptorVisibilities.PACKAGE_VISIBILITY
         }
-        var localPropertiesInClass = 0
 
-        irClass.transformChildrenVoid(object : IrElementTransformerVoidWithContext() {
-            override fun visitLocalDelegatedProperty(declaration: IrLocalDelegatedProperty): IrStatement {
-                localPropertyIndices[declaration.getter.symbol] = localPropertiesInClass++
-                return super.visitLocalDelegatedProperty(declaration)
-            }
+        val localProperties = mutableListOf<IrLocalDelegatedPropertySymbol>()
+        val localPropertyIndices = mutableMapOf<IrSymbol, Int>()
+        val isSynthetic = irClass.metadata !is MetadataSource.File && irClass.metadata !is MetadataSource.Class
+
+        fun localPropertyIndex(getter: IrSymbol): Int? =
+            localPropertyIndices[getter] ?: parent?.localPropertyIndex(getter)
+
+        fun localPropertyOwner(getter: IrSymbol): IrClass? =
+            if (getter in localPropertyIndices) irClass else parent?.localPropertyOwner(getter)
 
-            override fun visitPropertyReference(expression: IrPropertyReference): IrExpression =
-                cachedKProperty(expression)
+        fun rememberLocalProperty(property: IrLocalDelegatedProperty) {
+            // Prefer to attach metadata to non-synthetic classes, because it won't be serialized otherwise;
+            // if not possible, though, putting it right here will at least allow non-reflective uses.
+            val metadataOwner = generateSequence(this) { it.parent }.find { !it.isSynthetic } ?: this
+            metadataOwner.localPropertyIndices[property.getter.symbol] = metadataOwner.localProperties.size
+            metadataOwner.localProperties.add(property.symbol)
+        }
+    }
 
-            override fun visitLocalDelegatedPropertyReference(expression: IrLocalDelegatedPropertyReference): IrExpression =
-                cachedKProperty(expression)
+    private var currentClassData: ClassData? = null
```

Comment:
It'd be great to avoid mutable state here by passing this within the visitor's `data` variable. However that would also require using `IrElementTransformer` instead of `IrElementTransformerVoidWithContext` and passing `currentScope`/`currentDeclarationParent` as well, and I'm not sure that it'll be pretty, so I think it's not needed for this PR